internals/overlord/planstate/manager.go

```diff
@@ -0,0 +1,192 @@
+package planstate
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+// LabelExists is the error returned by AppendLayer when a layer with that
+// label already exists.
+type LabelExists struct {
+	Label string
+}
+
+func (e *LabelExists) Error() string {
+	return fmt.Sprintf("layer %q already exists", e.Label)
+}
+
+type PlanManager struct {
+	state     *state.State
+	runner    *state.TaskRunner
+	pebbleDir string
+
+	planLock     sync.Mutex
+	plan         *plan.Plan
+	planHandlers []PlanChangeFunc
+}
+
+func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string) (*PlanManager, error) {
+	manager := &PlanManager{
+		state:     s,
+		runner:    runner,
+		pebbleDir: pebbleDir,
+	}
+
+	return manager, nil
+}
+
+// Load reads plan layers from the pebble directory, combine and validate the
+// final plan, and finally notifies registered managers up the plan update.
+func (m *PlanManager) Load() error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	plan, err := plan.ReadDir(m.pebbleDir)
+	if err != nil {
+		return err
+	}
+	m.planChanged(plan)
+	return nil
+}
+
+// PlanChangeFunc is the type of function used by AddChangeListeners.
+type PlanChangeFunc func(p *plan.Plan)
```

Comment:
*nit*: `PlanChangedFunc` (it's a function that gets called when the plan changed, not a function to change the plan) - the function on `PlanManager` is also called `planChanged`

Reply:
Fixed. Thank you. Resolving.